packages/serialize/src/index.ts

```diff
@@ -296,34 +389,33 @@ function createStringFromObject(
 
 let labelPattern = /label:\s*([^\s;\n{]+)\s*(;|$)/g
 
-let sourceMapPattern
+let sourceMapPattern: RegExp | undefined
 if (process.env.NODE_ENV !== 'production') {
   sourceMapPattern =
     /\/\*#\ssourceMappingURL=data:application\/json;\S+\s+\*\//g
 }
 
 // this is the cursor for keyframes
 // keyframes are stored on the SerializedStyles object as a linked list
-let cursor
+let cursor: Cursor | undefined
 
-export const serializeStyles = function (
-  args /*: Array<Interpolation> */,
-  registered /*: RegisteredCache | void */,
-  mergedProps /*: void | Object */
-) /*: SerializedStyles */ {
+export const serializeStyles = function <Props>(
+  args: Array<TemplateStringsArray | Interpolation<Props>>,
+  registered: RegisteredCache,
+  mergedProps?: Props
+): SerializedStyles {
   if (
     args.length === 1 &&
-    typeof args[0] === 'object' &&
-    args[0] !== null &&
-    args[0].styles !== undefined
+    !Array.isArray(args[0]) &&
+    isSerializedStyles(args[0])
```

Comment:
Like you have mentioned - this package is quite specific as it might be hard for TS to understand all of the type refinements etc.

I would prefer for this PR to stick away from runtime changes as much as possible (if something is absolutely required then I'm open to considering it). This is a very hot path and I don't want to retest performance after those changes :wink:

In cases like this, it should be OK to just use type assertion within the given block, without introducing type predicates.

Reply:
Sure!
I came to the same conclusion after tests started breaking, I will revert these changes